python/sparkdl/graph/input.py

```diff
@@ -0,0 +1,317 @@
+#
+from __future__ import absolute_import, division, print_function
+
+import tensorflow as tf
+from tensorflow.core.protobuf import meta_graph_pb2  # pylint: disable=no-name-in-module
+
+import sparkdl.graph.utils as tfx
+
+__all__ = ["TFInputGraph"]
+
+# pylint: disable=invalid-name,wrong-spelling-in-comment,wrong-spelling-in-docstring
+
+
+class TFInputGraph(object):
+    """
+    An opaque object containing TensorFlow graph.
+    This object can be serialized.
+
+    .. note:: We recommend constructing this object using one of the class constructor methods.
+
+              - :py:meth:`fromGraph`
+              - :py:meth:`fromGraphDef`
+              - :py:meth:`fromCheckpoint`
+              - :py:meth:`fromCheckpointWithSignature`
+              - :py:meth:`fromSavedModel`
+              - :py:meth:`fromSavedModelWithSignature`
+
+
+    When the graph contains serving signatures in which a set of well-known names are associtated
+    with their corresponding raw tensor names in the graph, we extract and store them here.
+    For example, the TensorFlow saved model may contain the following structure,
+    so that end users can retrieve the the input tensor via `well_known_input_sig` and
+    the output tensor via `well_known_output_sig` without knowing the actual tensor names a priori.
+
+    .. code-block:: python
+
+        sigdef: {'well_known_prediction_signature':
+        inputs { key: "well_known_input_sig"
+          value {
+            name: "tnsrIn:0"
+            dtype: DT_DOUBLE
+            tensor_shape { dim { size: -1 } dim { size: 17 } }
+            }
+          }
+        outputs { key: "well_known_output_sig"
+          value {
+            name: "tnsrOut:0"
+            dtype: DT_DOUBLE
+            tensor_shape { dim { size: -1 } }
+            }
+        }}
+
+
+    In this case, the class will internally store the mapping from signature names to tensor names.
+
+    .. code-block:: python
+
+        {'well_known_input_sig': 'tnsrIn:0'}
+        {'well_known_output_sig': 'tnsrOut:0'}
+
+
+    :param graph_def: :py:obj:`tf.GraphDef`, a serializable object containing the topology and
+                       computation units of the TensorFlow graph. The graph object is prepared for
+                       inference, i.e. the variables are converted to constants and operations like
+                       BatchNormalization_ are converted to be independent of input batch.
+
+                       .. _BatchNormalization: https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization
+
+    :param input_tensor_name_from_signature: dict, signature key names mapped to tensor names.
+                                             Please see the example above.
+    :param output_tensor_name_from_signature: dict, signature key names mapped to tensor names
+                                              Please see the example above.
+    """
+
+
+    def __init__(self, graph_def, input_tensor_name_from_signature,
```

Comment:
I agree. Naming things is difficult, especially in this case. 
TF does not seem to have a proper name for `"well_known_input_sig"` yet. 
And `signature_def_key` refers to `"well_known_prediction_signature"` in our example. 
Maybe `input_signature_to_tensor_name` and `output_signature_to_tensor_name`?

Reply:
Yeah i think those are actually good, since it conveys the meaning that it's mapping from some signature thing to tensor, and really we don't need the user to know exactly what the keys are -- users aren't expected to use these directly.

Reply:
As long as the conversion methods (in part 4) make it clear what their inputs are.